### ISSUE TITLE
Refactor IndexKernel to be parameterized with Cholesky factor of correlation matrix if full rank

### ISSUE THREAD

**Issue by Balandat** [cornellius-gp/gpytorch#220]
See discussion here: https://github.com/cornellius-gp/gpytorch/pull/218#discussion_r207721880

**Comment by Balandat** [comment]
@hjmshi this is another case of optimizing subject to non-linear constraints. Basically, we want to parameterize a `n x n` correlation matrix `M` via its Cholesky factor `L` such that `M = LL^T`. Turns out that for `L` to make `M` a correlation matrix it needs to have positive diagonal and unit row length, that is, `\sum_j L_{ij}^2 =1` for all `i`. 

Alternatively, if one was to use the full correlation matrix as a variable, one could encode this as a bunch of linear equality constraints (all diagonal elements equal 1) together with a (symmetric) positive definiteness constraint on the matrix. Of course our object is not linear, so this is not an SDP...

**Comment by jacobrgardner** [comment]
@Balandat we don't actually need constrained optimization to maintain a Cholesky factor of at least a **covariance** matrix, we can just multiply the matrix rows by the sign of the diagonal, which doesn't change the result of performing M=LL'. 

See CholVariationalDistribution for an example of where we do just this with the variational covariance matrix.

**Comment by Balandat** [comment]
Yeah that makes sense. What I'm doing right now in #376 is effectively to parameterize a covariance matrix `M=LL'` via its Cholesky factors, and then normalize via `DMD`, where `D = diag(sqrt(diag(M))^-1)`, i.e. `D_ii = (\sum_j L_{ij}^2)^-1`. This introduces unnecessary degrees of freedom in the optimization though, so it might not be the best move.
